Add metrics command syntax

dnhatn opened the pull request:
This PR introduces a METRICS command in ESQL for timeseries indices.This only adds the metrics syntax and translates it into a pair of EsIndex and Aggregate logical plans. Subsequent pull requests will introduce new logical and physical plans for handling time-series aggregations.

Some examples:

- METRICS tsdb
- METRICS tsdb max(cpu) BY host
- METRICS pods load=avg(cpu), writes=max(rate(indexing_requests)) BY pod | SORT pod

dnhatn committed:
Add metrics command syntax

dnhatn commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4:
```diff
@@ -133,6 +134,14 @@ deprecated_metadata
+    ;
+
+statsOption
+    : STATS aggregates=fields? (BY grouping=fields)?
```
Compared to our document, I've added a `STATS` keyword between the indices and the aggregations. This clarifies the distinction between options and aggregations within the METRICS command. I'm find to remove it if needed. `METRICS pods OPTIONS "allow_no_indices"="true" STATS load=avg(cpu) BY pod | SORT pod`

kkrik-es replied:
Yeah I'm on the fence here. Do we have documentation on how OPTIONS is used in general?

Also, can we get an example with filtering through WHERE? Or, do you plan to include this in a follow-up?

dnhatn replied:
This is the index options for ES|QL: https://www.elastic.co/guide/en/elasticsearch/reference/master/esql-index-options.html

dnhatn replied:
++ I have pushed https://github.com/elastic/elasticsearch/pull/108115/commits/739e00af700dd685f9656b8ce44f04a260f12bfe to implement the original proposal.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

martijnvg commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/TimeSeriesIT.java:
```diff
@@ -37,6 +45,48 @@ public void testEmpty() {
+        }
+        List<String> sortedGroups = cpus.keySet().stream().sorted().toList();
+        client().admin().indices().prepareRefresh("pods").get();
+        try (EsqlQueryResponse resp = run("METRICS pods STATS load=avg(cpu) BY pod | SORT pod")) {
```
Shouldn't the `METRICS` command at least for now be protected by tsdb query pragma?

dnhatn replied:
The metrics command is only available in snapshot builds. I'm not sure whether we should require the time-series pragmas for it. It's easier to try it without the pragmas.

kkrik-es replied:
Agreed, we should start moving away from the pragma once we have a metrics cmd.

martijnvg replied:
I missed that the metrics command is only available in snapshot build. Only making it available in snapshot builds is good. The reason I mentioned this, as far as I understand the METRICS command isn't wired up yet in this PR to do time series aggregation. Additional changes are needed to do this.

astefan commented:
Also, what is the difference between STATS inside the METRICS command and the STATS command by itself?
Can they be used in the same command?

dnhatn committed:
Merge remote-tracking branch 'elastic/main' into metrics-syntax

dnhatn committed:
Defer options and metadata

dnhatn committed:
Merge remote-tracking branch 'elastic/main' into metrics-syntax

dnhatn committed:
Compact

dnhatn commented:
> Also, what is the difference between STATS inside the METRICS command and the STATS command by itself? Can they be used in the same command?

Thanks @astefan. At the same syntax level, they are the same. However, differences come after that:

- MetricsAggregate allows nested aggregation functions for the rate function. For example, `max(rate(requests))` is valid in MetricsAggregate but not in Aggregate.
- MetricsAggregate supports filtering with the time-series bucket function, which does not exist in Aggregate.
- MetricsAggregate will be executed in three stages instead of two stages like Aggregate.

This PR mainly focuses on syntax. I will introduce these plans in a follow-up.

dnhatn commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4:
```diff
@@ -133,6 +134,14 @@ deprecated_metadata
     ;
 
+metricsCommand
+    : METRICS fromIdentifier (COMMA fromIdentifier)* metadataOption? fromOptions? statsOption?
```
Sure, I added these tests in https://github.com/elastic/elasticsearch/pull/108115/commits/739e00af700dd685f9656b8ce44f04a260f12bfe.

dnhatn commented:
After discussing with Martijn, I have removed OPTIONS and METADATA from the METRICS command. As the command is only available in snapshot builds, we accept an incomplete syntax to iterating on planning, execution, and performance. This approach helps us to move quicker compared to focusing on syntax, which may change after the planning and execution phases.

@kkrik-es @astefan @bpintea It's ready again. Can you please take a look? Thank you!

dnhatn commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -428,3 +429,68 @@ SETTING_WS
+    : WS -> channel(HIDDEN)
+    ;
+
+mode CLOSING_METRICS_MODE;
```
I'm seeking suggestions on how to better handle mode transitions here.

kkrik-es reviewed (approved):
The syntax looks fine, I'll defer reviewing the parser changes to the experts.

dnhatn committed:
Merge remote-tracking branch 'elastic/main' into metrics-syntax

dnhatn committed:
Update antlr output

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -6,6 +6,7 @@ ENRICH : 'enrich'             -> pushMode(ENRICH_MODE);
 EVAL : 'eval'                 -> pushMode(EXPRESSION_MODE);
 EXPLAIN : 'explain'           -> pushMode(EXPLAIN_MODE);
 FROM : 'from'                 -> pushMode(FROM_MODE);
+METRICS : 'metrics'           -> pushMode(METRICS_MODE);
```
Nit: keep the alphabetical order of the commands.

dnhatn replied:
++, I pushed https://github.com/elastic/elasticsearch/pull/108115/commits/ba52c4419caf8d803f8b806a9084206b24e98a6b

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -424,3 +425,68 @@ SETTING_WS
+mode METRICS_MODE;
+METRICS_PIPE : PIPE -> type(PIPE), popMode;
+
+fragment METRICS_UNQUOTED_IDENTIFIER_PART
+    : ~[=`|,[\]/ \t\r\n]
+    | '/' ~[*/] // allow single / but not followed by another / or * which would start a comment
+    ;
```
No need to define this since it's already defined in the FROM command (line [196](https://github.com/elastic/elasticsearch/pull/108115/files#diff-e78771d02134981fe1563249a0fbbdabf95b40299b2086dd19fccef6e87f39c4R196))

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -424,3 +425,68 @@ SETTING_WS
+    : METRICS_UNQUOTED_IDENTIFIER_PART+ -> popMode, pushMode(CLOSING_METRICS_MODE)
+    ;
+
+METRICS_QUOTED_IDENTIFIER
+    : QUOTED_IDENTIFIER -> type(QUOTED_IDENTIFIER), popMode, pushMode(CLOSING_METRICS_MODE)
+    ;
```
Please remove this for now, as we might opt for a different syntax/parsing for quoted locations.

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -424,3 +425,68 @@ SETTING_WS
+    ;
+
+METRICS_UNQUOTED_IDENTIFIER
+    : METRICS_UNQUOTED_IDENTIFIER_PART+ -> popMode, pushMode(CLOSING_METRICS_MODE)
```
why us a separate mode and not follow the STATS implementation for now? This will simplify your grammar and remove some subtleties around mode management.

`FROM_UNQUOTED_IDENTIFIER -> type(FROM_QUOTED_IDENTIFIER)`

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4:
```diff
@@ -132,6 +133,15 @@ deprecated_metadata
+    : METRICS metricsIdentifier (COMMA metricsIdentifier)* aggregates=fields? (BY grouping=fields)?
+    ;
+
+metricsIdentifier
```
Reuse `fromIdentifier` or better yet define a rule to be shared across both commands:

```
indexDefinitions
   : indexDefinition (COMMA indexDefinition)*
   ;

indexDefinition
  : FROM_UNQUOTED_IDENTIFIER
  ;

...
metricsCommand METRICS indexDefinitions aggregates=field? (BY grouping=fields)?
```

dnhatn replied:
++ I combined these rules in https://github.com/elastic/elasticsearch/pull/108115/commits/7740ed443c2aef2a0360257370c179300fbcacea.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java:
```diff
@@ -427,5 +437,21 @@ private static Tuple<Mode, String> parsePolicyName(Token policyToken) {
+        } else {
+            stats = new Stats(List.of(), List.of());
+        }
+        return new UnresolvedMetrics(source, table, null, stats.groupings, stats.aggregates());
```
Based on the resolution rule, UnresolvedMetrics doesn't carry any extra information - instead of introducing this temporary object (which doesn't survive the analysis phase) simply return the aggregate plus unresolved relation (which will get resolved automatically):

```
visitMetricsCommand(...) {
  ...
val relation = new EsqlUnresolvedRelation(source, index, ...);
return input -> new EsqlAggregate(source(ctx), relation, stats.grouping, stats.aggregates);
} 
```

dnhatn replied:
We need to introduce this relation for the METRICS command so that we can carry the filter and map it to the TimeSeriesSortedSourceOperatorFactory. However, it's better to remove it now from this PR to avoid confusion. I've removed it in https://github.com/elastic/elasticsearch/pull/108115/commits/435ab0082cccca3a28f3178d829dab4c91564201.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/UnresolvedMetrics.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Unresolved logical plan from METRICS command
+ */
+public final class UnresolvedMetrics extends EsqlUnresolvedRelation {
```
See my previous comment - instead of defining a new class (which is short lived), use the parser to create the EsqlUnresolvedRelation which requests _tsid metaata and timestamp field.

costin reviewed (commented):
Two main comments:
- use the stats definition and don't go for nested modes, these are not needed.
- create the dedicated esqlaggreate+esrelation inside the parser - this avoids creating UnresolvedMetrics and modifying the existing resolution rules.

dnhatn committed:
Remove UnresolvedMetrics

dnhatn committed:
Alphabet

dnhatn committed:
index identifier

dnhatn commented:
> use the stats definition and don't go for nested modes, these are not needed.

Thanks Costin. I think I have addressed all your comments except this. Here, we need to move from the IndexIdentifier mode to the expression mode. Do you have any suggestion to solve this without introducing two modes here?

dnhatn committed:
Merge remote-tracking branch 'elastic/main' into metrics-syntax

costin reviewed (approved):
The modes are not necessary however since this is WIP, LGTM

dnhatn committed:
Add TODO

dnhatn committed:
Merge branch 'main' into metrics-syntax

dnhatn commented:
Thanks everyone for reviews and feedback.
